fix(core): distinguish injection tokens based on their value type

crisbeto opened the pull request:
Currently the generic type of `InjectionToken` isn't used anywhere which means that the compiler won't flag the following as an error:

```
const foo = new InjectionToken<number>('foo');
const bar: InjectionToken<string> = foo; // Should be an error, but it's not.
```

These changes add a private property to allow the compiler to distinguish between the two.

Fixes #46815.

crisbeto commented:
This [passed a TGP](http://test/OCL:474767232:BASE:474780569:1663326119878:91a1f147) so I'm assuming that it won't be particularly breaking externally either. I'll bump it up to `major` just in case.

crisbeto commented on packages/core/testing/src/test_bed.ts:
```diff
@@ -553,10 +553,11 @@ export class TestBedImpl implements TestBed {
     // TODO: Don't cast as `InjectionToken<boolean>`, proper type is boolean[]
-    const noNgZone = this.inject(ComponentFixtureNoNgZone as InjectionToken<boolean>, false);
+    const noNgZone =
+        this.inject(ComponentFixtureNoNgZone as unknown as InjectionToken<boolean>, false);
```
I went for the `as unknown` here, since there's a TODO above mentioning that the injection token isn't typed correctly.

crisbeto commented on packages/core/src/di/injection_token.ts:
```diff
@@ -87,7 +100,7 @@ export class InjectionToken<T> {
    */
   get multi(): InjectionToken<Array<T>> {
-    return this as InjectionToken<Array<T>>;
+    return this as unknown as InjectionToken<Array<T>>;
```
This change only broke in g3. I assume it has something to do with the fact that g3 is still on TS 4.7.

AndrewKushnir replied:
Could you please leave a quick comment in the code here to mention that? It'd help find this place later and try to remove the `unknown` once g3 is on a later version of TS.

devversion commented on packages/core/src/di/injection_token.ts:
```diff
@@ -58,6 +58,19 @@ export class InjectionToken<T> {
+   *
+   * @internal
+   */
+  private value!: T;
```
Have you tested how this works when shipped to NPM?

1. The `private` modifier causes the type `T` to be omitted in the shipped `d.t.s` IIRC ([see](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAYwDYEMDOa4A0A8AVAPjgG8AoOCuMKASwDcUZgqBCALjjwG5SBfUoA))
2. We use `stripInternal`- so the whole field would be omitted for our consumers ([see](https://github.com/bazelbuild/rules_nodejs/blob/stable/packages/concatjs/internal/common/tsconfig.bzl#L279))

crisbeto replied:
I haven't tried packaging and testing it, but I assumed it would be fine since we have other places with these sorts of branding properties.

AndrewKushnir replied:
+1 to test this after packaging and/or add a check in one of the apps within `integration` folder (where apps work with built packages).

crisbeto replied:
I ended up making it a public `__brand__` property. We appear to be using this pattern in a few other places.

JoostK replied:
A `private` field is not stripped from `.d.ts` files, Paul's playground link shows the following .d.ts output:

```ts
export declare class X<T> {
    private p;
}
```

so it should be possible to make it `private`, but not `@internal`.

devversion replied:
yes, the field is not stripped- but the type.

JoostK replied:
Oh, how did I miss that 🤦‍♂️

crisbeto committed:
fix(core): distinguish injection tokens based on their value type

Currently the generic type of `InjectionToken` isn't used anywhere which means that the compiler won't flag the following as an error:

```
const foo = new InjectionToken<number>('foo');
const bar: InjectionToken<string> = foo; // Should be an error, but it's not.
```

These changes add a private property to allow the compiler to distinguish between the two.

Fixes #46815.

AndrewKushnir reviewed (approved):
👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

jessicajaniuk reviewed (approved):
reviewed-for: public-api

Harpush commented:
@crisbeto any chance for it to land in v16?

crisbeto commented:
Likely not. I haven't had the time to migrate the internal codebase to avoid the breakages.

thePunderWoman reviewed (approved):
reviewed-for: public-api
